add thanos-ruler operator

pgier opened the pull request:
This change adds a new CRD and associated operator/controller for ThanosRuler instances.
This allows ThanosRuler instances to be deployed according to the
configuration in the ThanosRuler objects.  A rule selector can be used
to select PrometheusRule objects which are combined into a ConfigMap and
mounted into the associated ThanosRuler pod.

brancz commented on Documentation/api.md:
```diff
@@ -644,3 +648,61 @@ ThanosSpec defines parameters for a Prometheus server within a Thanos deployment
+
+| Field | Description | Scheme | Required |
+| ----- | ----------- | ------ | -------- |
+| podMetadata | Standard object’s metadata. More info: https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#metadata Metadata Labels and Annotations gets propagated to the prometheus pods. | *[metav1.ObjectMeta](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#objectmeta-v1-meta) | false |
```
`paused` seems to be missing in the spec

pgier replied:
I've added this now.

brancz commented on pkg/thanos/rules.go:
```diff
@@ -0,0 +1,334 @@
+// large buffer.
+var maxConfigMapDataSize = int(float64(v1.MaxSecretSize) * 0.5)
+
+func (o *Operator) createOrUpdateRuleConfigMaps(t *monitoringv1.ThanosRuler) ([]string, error) {
```
How come we redefine this here? Is this not reusable from the Prometheus flow that does the same thing? I'd expect the Thanos ruler and Prometheus to use shared code for this (maybe this is simply a todo, in which case ignore this comment).

pgier replied:
I didn't reuse it directly at first because it uses several fields from the Prometheus spec, but it seems the logic is the same other than the names of things.  So I think you are right that it should be reusable after refactoring to make the original functions a bit more generic.

brancz replied:
Yes I assume something to that regard, makes perfect sense!

brancz replied:
I would still like to see this before we merge.

brancz commented:
Very nice start!

metalmatze commented:
Running the operator with this version and no ThanosRule resource available I sometimes get this panic:
```
E0114 20:56:06.314960       1 runtime.go:78] Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(0x13fe480), concrete:(*runtime._type)(0x158bc00), asserted:(*runtime._type)(0x158c180), missingMethod:""} (interface conversion: interface {} is *v1.PrometheusRule, not *v1.ThanosRuler)
```

metalmatze commented on pkg/apis/monitoring/v1/thanos_types.go:
```diff
@@ -0,0 +1,134 @@
+	ListenLocal bool `json:"listenLocal,omitempty"`
+	// QueryEndpoints defines Thanos querier endpoints from which to query metrics.
+	// Maps to the --query flag of thanos ruler.
+	QueryEndpoints []string `json:"queryEndpoints"`
```
This seems to be a required field right now.

I'm just guessing but maybe this makes it optional
```suggestion
	QueryEndpoints *[]string `json:"queryEndpoints"`
```

pgier replied:
It's required to have some value when starting the container.  This field just translates to `--query` arguments to the thanos binary, and at least one is required when running `thanos rule`.  I guess this is because without any query endpoint you could never evaluate any rules.

simonpasquier replied:
Correct though it has changed a bit with the addition of the `--query-config*` flags which supersede `--query`.

https://github.com/thanos-io/thanos/blob/46a97fdc9765bbe7eb9d75d801abf186015e1b01/cmd/thanos/rule.go#L148-L150

metalmatze commented:
The StatefulSet that gets created is called `thanos-metalmatze` in my case, with `metalmatze` as name. This should be `thanos-ruler-metalmatze`. :blush:

metalmatze commented:
Sadly I had to revert because of the panic above, but I'm just 2 CLI commands away from giving it another try on my cluster. Let me know! :+1:

pgier commented:
@metalmatze Thanks for the feedback!  I believe both issues are addressed now, I changed the prefix of the statefulset to `thanos-ruler` instead of `thanos`, and I believe the type mismatch bug is fixed now

metalmatze commented:
I've tested the latest changes and can indeed see the updated statefulset name. Very cool! :)

Sadly, I still get this panic in the Operator:
```
panic: interface conversion: interface {} is *v1.ThanosRuler, not *v1.Prometheus

goroutine 252 [running]:
github.com/coreos/prometheus-operator/pkg/thanos.(*thanosRulerCollector).Collect(0xc0003ef680, 0xc000581560)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/pkg/thanos/collector.go:51 +0xd3
github.com/prometheus/client_golang/prometheus.(*wrappingCollector).Collect.func1(0xc0004078f0, 0xc000581560)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/github.com/prometheus/client_golang/prometheus/wrap.go:115 +0x3d
created by github.com/prometheus/client_golang/prometheus.(*wrappingCollector).Collect
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/github.com/prometheus/client_golang/prometheus/wrap.go:114 +0x6b
```

pgier commented:
Ok, looks like the error you are seeing is similar but a slightly different copy/paste issue from the one I fixed earlier.  It should be fixed  now, hopefully for real this time.

metalmatze commented:
Still one last panic:
```
E0116 13:43:43.008137       1 runtime.go:78] Observed a panic: &runtime.TypeAssertionError{_interface:(*runtime._type)(0x13fe480), concrete:(*runtime._type)(0x158bc00), asserted:(*runtime._type)(0x158c180), missingMethod:""} (interface conversion: interface {} is *v1.PrometheusRule, not *v1.ThanosRuler)
goroutine 292 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1431ea0, 0xc00083f9b0)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1431ea0, 0xc00083f9b0)
	/usr/lib/go/src/runtime/panic.go:679 +0x1b2
github.com/coreos/prometheus-operator/pkg/thanos.(*Operator).handleThanosRulerUpdate(0xc000426380, 0x158bc00, 0xc0001af2c0, 0x158bc00, 0xc0001af2c0)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/pkg/thanos/operator.go:328 +0x3ad
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/client-go/tools/cache/controller.go:205
k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x1, 0x0, 0x0)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:656 +0x188
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0, 0xc00020edd8, 0x0, 0xc0008d1ab8)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:292 +0x51
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:652 +0x79
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00041b740)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc00020ef40, 0xdf8475800, 0x0, 0xc00007ca01, 0xc0001591a0)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*processorListener).run(0xc00073e280)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:650 +0x9b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc000456d80, 0xc000344ff0)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x59
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62
panic: interface conversion: interface {} is *v1.PrometheusRule, not *v1.ThanosRuler [recovered]
	panic: interface conversion: interface {} is *v1.PrometheusRule, not *v1.ThanosRuler

goroutine 292 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x105
panic(0x1431ea0, 0xc00083f9b0)
	/usr/lib/go/src/runtime/panic.go:679 +0x1b2
github.com/coreos/prometheus-operator/pkg/thanos.(*Operator).handleThanosRulerUpdate(0xc000426380, 0x158bc00, 0xc0001af2c0, 0x158bc00, 0xc0001af2c0)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/pkg/thanos/operator.go:328 +0x3ad
k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/client-go/tools/cache/controller.go:205
k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0x1, 0x0, 0x0)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:656 +0x188
k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0, 0xc000b91dd8, 0x0, 0xc0008d1ab8)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:292 +0x51
k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:652 +0x79
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc00041b740)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5e
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000b91f40, 0xdf8475800, 0x0, 0xc00007ca01, 0xc0001591a0)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(...)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
k8s.io/client-go/tools/cache.(*processorListener).run(0xc00073e280)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/client-go/tools/cache/shared_informer.go:650 +0x9b
k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc000456d80, 0xc000344ff0)
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x59
created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/home/metalmatze/src/github.com/coreos/prometheus-operator/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62
```

I started that last evening with the latest changes. 
The StatefulSet is created just fine and the Operator can make some progress before this happens.

pgier commented:
@metalmatze rebased and another update to hopefully fix the remaining panic and also correctly update configmaps used by thanos-ruler when there are changes to the associated prometheusrules.

simonpasquier commented on pkg/thanos/operator.go:
```diff
@@ -0,0 +1,774 @@
+
+	if ok && sErr.ErrStatus.Code == 422 && sErr.ErrStatus.Reason == metav1.StatusReasonInvalid {
+		o.metrics.StsDeleteCreateCounter().Inc()
+		level.Info(o.logger).Log("msg", "resolving illegal update of Alertmanager StatefulSet", "details", sErr.ErrStatus.Details)
```
s/Alertmanager/Thanos/

pgier replied:
Done

simonpasquier commented on pkg/thanos/operator.go:
```diff
@@ -0,0 +1,774 @@
+
+	o.thanosRulerInf = cache.NewSharedIndexInformer(
+		o.metrics.NewInstrumentedListerWatcher(
+			listwatch.MultiNamespaceListerWatcher(o.logger, o.config.Namespaces.AlertmanagerAllowList, o.config.Namespaces.DenyList, func(namespace string) cache.ListerWatcher {
```
s/AlertmanagerAllowList/ThanosAllowList/ ?

pgier replied:
Done

pgier commented on Documentation/api.md:
```diff
@@ -644,3 +648,64 @@ ThanosSpec defines parameters for a Prometheus server within a Thanos deployment
+| ruleNamespaceSelector | Namespaces to be selected for Rules discovery. If unspecified, only the same namespace as the ThanosRuler object is in is used. | *[metav1.LabelSelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#labelselector-v1-meta) | false |
+| enforcedNamespaceLabel | EnforcedNamespaceLabel enforces adding a namespace label of origin for each alert and metric that is user created. The label value will always be the namespace of the object that is being created. | string | false |
+| logLevel | Log level for Alertmanager to be configured with. | string | false |
+| logFormat | Log format for Alertmanager to be configured with. | string | false |
```
Done

pgier commented on Documentation/api.md:
```diff
@@ -644,3 +648,64 @@ ThanosSpec defines parameters for a Prometheus server within a Thanos deployment
+| ruleSelector | A label selector to select which PrometheusRules to mount for alerting and recording. | *[metav1.LabelSelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#labelselector-v1-meta) | false |
+| ruleNamespaceSelector | Namespaces to be selected for Rules discovery. If unspecified, only the same namespace as the ThanosRuler object is in is used. | *[metav1.LabelSelector](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.11/#labelselector-v1-meta) | false |
+| enforcedNamespaceLabel | EnforcedNamespaceLabel enforces adding a namespace label of origin for each alert and metric that is user created. The label value will always be the namespace of the object that is being created. | string | false |
+| logLevel | Log level for Alertmanager to be configured with. | string | false |
```
Done

metalmatze commented:
Made some more progress on my setup with the thanos-ruler. 
First I noticed a probable races between the ConfigMap and the StatefulSet being created. In 10 re-deploys I had roughly 2 occurrences where the ConfigMap didn't yet exist upon Pod creation.
We should try make sure the ConfigMap is there before creating the StatefulSet.

I also noticed that there's currently no Service created for this Ruler, where for Prometheus and Alertmanager it gets created by the Operator (I'm pretty sure it's not kube-prometheus, too late in Berlin and I am too lazy to double check right now, sorry).

pgier commented:
I've rebased and squashed the commits.  Added an alertmanagers.config option as @simonpasquier suggested.  I'll wait for Thanos v0.11.0 before adding query.config.  I've also added Service creation as @metalmatze suggested.

@metalmatze Regarding the race condition, I didn't really see any difference between the new Thanos operator and the Prometheus operator, I'm guessing the issue just doesn't show up in Prometheus stateful sets because there are few more steps between creating the configmap and the stateful set such as creating secrets.  Hopefully, the addition of the service creation between configmap and statefulset helps a bit at least for now, and I'll try to figure out a better way to ensure that the configmap is available when needed.

pgier commented:
I made a few more updates and realized there was no configmap informer set up in the thanos operator. so that might have been the cause of the race issues.  @metalmatze everything should be ready for you to test again.

simonpasquier commented on pkg/thanos/statefulset.go:
```diff
@@ -0,0 +1,337 @@
+	}
+
+	if tr.Spec.AlertManagersConfig != nil {
+		trCLIArgs = append(trCLIArgs, "--objstore.config=$(ALERTMANAGERS_CONFIG)")
```
`--alertmanagers.config` instead of `--objstore.config`

pgier replied:
Done

metalmatze commented on Documentation/thanos.md:
````diff
@@ -95,7 +87,34 @@ This will attach Thanos sidecar that will backup all _new blocks_ that Prometheu
+    - dnssrv+_http._tcp.my-thanos-querier.monitoring.svc.cluster.local
+```
+
+The recording and alerting rules used by a `ThanosRuler` component, are configured using the same `PrometheusRule` objects which are used by Prometheus.  In the given example, the rules contained in any `PrometheusRule` object which match the label `role=my-thanos-rules` will be added to the Thanos Ruler POD.
````
We should add it to the `PrometheusRule` and then ignore it for Prometheus as you said. The entire reasoning behind putting the ThanosRuler in here was to reuse `PrometheusRules` for Thanos. :blush:

pgier replied:
Done.

pgier commented on Documentation/thanos.md:
```diff
@@ -95,7 +87,34 @@ This will attach Thanos sidecar that will backup all _new blocks_ that Prometheu
+## Thanos Ruler
+
+The [Thanos Ruler](https://github.com/thanos-io/thanos/blob/master/docs/components/rule.md) component allows recording and alerting rules to be processed across
+multiple Promtheus instances.  The Thanos Ruler queries metrics from a Thanos Querier.  It is therefore recommended to create at least one ThanosQuerier before creating a ThanosRuler.  To deploy a `ThanosRuler` instance, at least one
```
Update the doc

pgier commented on Documentation/thanos.md:
```diff
@@ -5,29 +5,21 @@ _Note: This guide is valid for Prometheus Operator v0.28+ and Thanos v0.2+ and a
 
-Thanos includes other components like queriers or rulers. To get the advantage of object storage it also requires compactors and stores.
+Prometheus Operator operates `Prometheus` and optionally `ThanosRuler` components.
+Other Thanos compents, such as the querier and store gateway, must be configured
```
done

brancz commented on pkg/apis/monitoring/v1/types.go:
```diff
@@ -813,11 +813,15 @@ type PrometheusRuleSpec struct {
 
 // RuleGroup is a list of sequentially evaluated recording and alerting rules.
+// Note: PartialResponseStrategy is only used by ThanosRuler and will
+// be ignored by Prometheus instances.  Valid values for this field are 'warn'
```
I don't see us ignoring the field. Are we sure Prometheus wouldn't refuse to load the rules?

simonpasquier replied:
Indeed Prometheus will reject rules with the `partial_response_strategy` field. But AFAICT the field is emptied when generating rules for Prometheus: https://github.com/coreos/prometheus-operator/blob/ca852a3d20f8051c62383d3858e9dca622ddf7ac/pkg/prometheus/rules.go#L213

brancz replied:
ack

brancz replied:
this seems somewhat risky, I would want us to make sure that we create a copy of the rule before we set something like this, as we are modifying the item in the cache with this line

brancz replied:
cc @pgier ^ we should do this in a follow up

metalmatze commented:
From a user perspective I think we can merge it for now and then do some follow up PRs in case we encounter more. The ruler is now running in my personal cluster for a few days! :tada:

brancz commented:
Ok as @metalmatze said, let's merge and continue to iterate.

s-urbaniak commented:
Thank you @brancz @metalmatze @pgier ! :raised_hands:

benjaminhuo commented:
Great feature 👍
I'm wondering if enabling Thanos ruler and querier require changing prometheus chunk to 2h?

Thanks
Ben
